lib/rules/prop-types.js

```diff
@@ -569,125 +611,210 @@ module.exports = {
+        if (!parentIsDestructured(node)) {
+          return void 0;
+        }
+
+        // if the parent is using destructuring, mark prop types for its properties
+        const properties = node.parent.id.properties;
+        markPropTypesInProperties(properties, usedPropTypes);
+        return void 0;
+      }
+
+      // this.props vs props
+      const startingNode = isThisProps ? node.parent : node;
+      const directlyOffOfProps = startingNode.object &&
+        (startingNode.object.name === 'props' || startingNode.object.name === 'nextProps');
+
+      // if this member expression isn't coming off props in some way
+      // don't search for a parent usage
+      if (directlyOffOfProps || isThisProps) {
+        return {
+          startingNode: startingNode,
+          parentUsage: void 0
+        };
+      }
+
+      // let { a } = this.props
+      // <div>{a.b}</div>
+      const parentUsage = usedPropTypes.find(
+        propType => propType.name === node.object.name
+      );
+
+      // if we're not directly off of props
+      // and there isn't a parent usage, skip this node
+      if (!parentUsage) {
+        return void 0;
+      }
+      return {
+        startingNode: startingNode,
+        parentUsage: parentUsage
+      };
+    }
+
+
+    /**
+     * Mark a node as using a prop type
+     * @param {ASTNode} node The AST node being marked.
+     */
+    function markPropTypesAsUsed(node) {
+      const component = components.get(utils.getParentComponent());
+      const usedPropTypes = (component && component.usedPropTypes || []).slice();
       switch (node.type) {
         case 'MemberExpression':
-          name = getPropertyName(node);
-          if (name) {
-            allNames = parentNames.concat(name);
-            if (node.parent.type === 'MemberExpression') {
-              markPropTypesAsUsed(node.parent, allNames);
-            }
-            // Do not mark computed props as used.
-            type = name !== '__COMPUTED_PROP__' ? 'direct' : null;
-          } else if (
-            node.parent.id &&
-            node.parent.id.properties &&
-            node.parent.id.properties.length &&
-            getKeyValue(node.parent.id.properties[0])
-          ) {
-            type = 'destructuring';
-            properties = node.parent.id.properties;
+          const processedExpression = getProcessedMemberExpression(node, usedPropTypes);
+          if (!processedExpression) {
+            break;
           }
+          const startingNames = processedExpression.parentUsage ? processedExpression.parentUsage.allNames : [];
+          traverseMemberExpression(processedExpression.startingNode, (propertyName, currentNames, currentNode) => {
+            // Ignore Object methods
+            if (Object.prototype[propertyName] || propertyName === '__COMPUTED_PROP__') {
+              return;
+            }
+
+            // Ignore Array methods
+            if (Array.prototype[propertyName]) {
```

Comment:
let's be a bit stricter here and check for `typeof === 'function'`? Also, this will include Object.prototype methods.

Also, what about when a method is shimmed/polyfilled onto Array.prototype in their code, but not in the node process that eslint is running in? I'm not sure this is a safe way to check it.